pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,272 @@
+		defer cancel()
+		localCtx = user.InjectOrgID(localCtx, userID)
+		if sp := opentracing.SpanFromContext(r.Context()); sp != nil {
+			localCtx = opentracing.ContextWithSpan(localCtx, sp)
+		}
+		// Get clientIP(s) from Context and add it to localCtx
+		localCtx = util.AddSourceIPsToOutgoingContext(localCtx, source)
+
+		resp, err := d.doRequest(localCtx, am, &httpgrpc.HTTPRequest{
+			Method:  r.Method,
+			Url:     r.RequestURI,
+			Body:    body,
+			Headers: grpcHeaders,
+		})
+		if err != nil {
+			d.amSendFailures.WithLabelValues(am.Addr).Inc()
+			return err
+		}
+
+		if resp.HttpResponse.Code/100 != 2 {
+			return httpgrpc.ErrorFromHTTPResponse(resp.HttpResponse)
+		}
+
+		select {
+		case firstSuccessfulResponse <- resp:
+		default:
+		}
+
+		return nil
+	}, func() {})
+
+	if err != nil {
+		d.respondFromError(err, w, logger)
+	}
+
+	select {
+	case resp := <-firstSuccessfulResponse:
+		d.respondFromGRPCHTTPResponse(w, resp.HttpResponse)
+	default:
+		// This should not happen, but we have this default case to prevent
+		// and bugs deadlocking at this point.
+		level.Warn(logger).Log("msg", "first successful response not available on nil error")
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+func (d *Distributor) doRead(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	key := shardByUser(userID)
+	replicationSet, err := d.alertmanagerRing.Get(key, RingOp, nil, nil, nil)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get replication set from the ring", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	req, err := server.HTTPRequest(r)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to get grpc request from http request", "err", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(r.Context(), d.cfg.AlertmanagerClient.RemoteTimeout)
+	defer cancel()
+	// Until we have a mechanism to combine the results from multiple alertmanagers,
+	// we forward the request to only only of the alertmanagers.
+	amDesc := replicationSet.Ingesters[0]
+	d.amSends.WithLabelValues(amDesc.Addr).Inc()
+	resp, err := d.doRequest(ctx, amDesc, req)
+	if err != nil {
+		d.amSendFailures.WithLabelValues(amDesc.Addr).Inc()
+		d.respondFromError(err, w, logger)
+		return
+	}
+
+	http.Error(w, string(resp.HttpResponse.Body), int(resp.HttpResponse.Code))
+}
+
+func (d *Distributor) respondFromError(err error, w http.ResponseWriter, logger log.Logger) {
+	httpResp, ok := httpgrpc.HTTPResponseFromError(errors.Cause(err))
+	if !ok {
+		level.Error(logger).Log("msg", "Failed to process the request to the alertmanager", "err", err)
+		http.Error(w, "Failed to process the request to the alertmanager", http.StatusInternalServerError)
+		return
+	}
+	d.respondFromGRPCHTTPResponse(w, httpResp)
+}
+
+func (d *Distributor) respondFromGRPCHTTPResponse(w http.ResponseWriter, httpResp *httpgrpc.HTTPResponse) {
```

Comment:
It's called "httpgrpc" not "grpchttp".

```suggestion
func (d *Distributor) respondFromHTTPGRPCResponse(w http.ResponseWriter, httpResp *httpgrpc.HTTPResponse) {
```